src/tls/openssl/tls.c

```diff
@@ -337,55 +339,63 @@ int tls_set_selfsigned(struct tls *tls, const char *cn)
 /**
  * Set the certificate and private key on a TLS context
  *
- * @param tls TLS Context
- * @param pem Certificate and private key in PEM format
- * @param len Length of PEM string
+ * @param tls      TLS Context
+ * @param cert     Certificate in PEM format
+ * @param len_cert Length of certificate PEM string
+ * @param key      Private key in PEM format, will be read from cert if NULL
+ * @param len_key  Length of private key PEM string
  *
  * @return 0 if success, otherwise errorcode
  */
-int tls_set_certificate(struct tls *tls, const char *pem, size_t len)
+int tls_set_certificate_pem(struct tls *tls, const char *cert, size_t len_cert,
+		const char *key, size_t len_key)
 {
 	BIO *bio = NULL, *kbio = NULL;
-	X509 *cert = NULL;
-	RSA *rsa = NULL;
+	X509 *x509 = NULL;
+	EVP_PKEY *pkey = NULL;
 	int r, err = ENOMEM;
 
-	if (!tls || !pem || !len)
+	if (!tls || !cert || !len_cert || (key && !len_key))
 		return EINVAL;
 
-	bio  = BIO_new_mem_buf((char *)pem, (int)len);
-	kbio = BIO_new_mem_buf((char *)pem, (int)len);
+	if (!key) {
+		key = cert;
+		len_key = len_cert;
+	}
+
+	bio = BIO_new_mem_buf((char *)cert, (int)len_cert);
+	kbio = BIO_new_mem_buf((char *)key, (int)len_key);
 	if (!bio || !kbio)
 		goto out;
 
-	cert = PEM_read_bio_X509(bio, NULL, 0, NULL);
-	rsa = PEM_read_bio_RSAPrivateKey(kbio, NULL, 0, NULL);
-	if (!cert || !rsa)
+	x509 = PEM_read_bio_X509(bio, NULL, 0, NULL);
+	pkey = PEM_read_bio_PrivateKey(kbio, NULL, 0, NULL);
+	if (!x509 || !pkey)
 		goto out;
 
-	r = SSL_CTX_use_certificate(tls->ctx, cert);
+	r = SSL_CTX_use_certificate(tls->ctx, x509);
 	if (r != 1)
 		goto out;
 
-	r = SSL_CTX_use_RSAPrivateKey(tls->ctx, rsa);
+	r = SSL_CTX_use_PrivateKey(tls->ctx, pkey);
 	if (r != 1) {
-		DEBUG_WARNING("set_certificate: use_RSAPrivateKey failed\n");
+		DEBUG_WARNING("set_certificate_pem: use_PrivateKey failed\n");
 		goto out;
 	}
 
 	if (tls->cert)
 		X509_free(tls->cert);
 
-	tls->cert = cert;
-	cert = NULL;
+	tls->cert = x509;
+	x509 = NULL;
 
 	err = 0;
 
- out:
-	if (cert)
-		X509_free(cert);
-	if (rsa)
-		RSA_free(rsa);
+out:
```

Comment:
Oh... I didn't even notice that. Out of curiosity: Any particular reason why you're adding a space before a label?